OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala

```diff
@@ -374,140 +315,144 @@ class AndroidICCAnalysis(val project: SomeProject) extends FPCFAnalysis {
-                    case PCAndInstruction(pc, in: INVOKEVIRTUAL) ⇒
-                        val index = tacCode.pcToIndex(pc)
-                        if (index > -1) {
-                            val methodCall = tacCode.stmts(index)
-                            if (methodCall.isVirtualMethodCall) {
-                                var i = 0
-                                methodCall.asVirtualMethodCall.descriptor.parameterTypes.foreach { p ⇒
-                                    if (p.isObjectType) {
-                                        val ot = p.toString.replaceAll(otRegex, "")
-                                        //only interfaces defined in 'callbacklist' are handled as callback interfaces and dynamically searching for callback interfaces
-                                        if (callbackList.contains(ot) ||
-                                            (ot.startsWith("android") && ot.endsWith("Listener") && ot.substring(ot.lastIndexOf("$") + 1).startsWith("On"))) {
-                                            val par = methodCall.asVirtualMethodCall.params(i).asVar.value.verificationTypeInfo
-                                            if (par.isObjectVariableInfo) {
-                                                val cf = project.classFile(par.asObjectVariableInfo.clazz.asObjectType)
-                                                if (cf.isDefined) {
-                                                    val methods = cf.get.methods.filter(f ⇒ f.name != init && f.name != "<clinit>")
-                                                    methods.foreach { method ⇒
-                                                        if (!callbackMethods.contains(method)) callbackMethods += method
-                                                    }
+            //find callbacks
+            val callbackMethods = ListBuffer.empty[Method]
+            body.collectWithIndex {
+                case PCAndInstruction(pc, in: INVOKEVIRTUAL) ⇒
+                    val index = tacCode.pcToIndex(pc)
+                    if (index > -1) {
+                        val methodCall = tacCode.stmts(index)
+                        if (methodCall.isVirtualMethodCall) {
+                            var i = 0
+                            methodCall.asVirtualMethodCall.descriptor.parameterTypes.foreach { p ⇒
+                                if (p.isObjectType) {
+                                    val ot = p.toString.replaceAll(otRegex, "")
+                                    //only interfaces defined in 'callbacklist' are handled as callback interfaces and dynamically searching for callback interfaces
+                                    if (callbackList.contains(ot) ||
+                                        (ot.startsWith("android") && ot.endsWith("Listener") && ot.substring(ot.lastIndexOf("$") + 1).startsWith("On"))) {
+                                        val par = methodCall.asVirtualMethodCall.params(i).asVar.value.verificationTypeInfo
```

Comment:
I am using the verificationTypeInfo to get the ObjectType of the parameter(i) of the methodCall, because i need to get the classFile. Is there a more elegant way to access the classFile of this parameter?

Reply:
As I said, value.asReferenceValue.upperTypeBound might be the way to do it.